src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+#include "pandaproxy/logger.h"
+
+namespace pandaproxy {
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg, size_t max_size, std::chrono::milliseconds keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{std::move(keep_alive)} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(user.name);
+        cfg.scram_password.set_value(user.pass);
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+client_ptr kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    auto it_hash = inner_hash.find(user.name);
```

Comment:
we are caching by user.name here which makes sense. what happens if a client is created from a client that provided the wrong password and then the client changes their password? would they have to wait for the client to expire out of the cache before the fixed password would create a new client and work?

Reply:
Excellent use-case that I did not account for. You are correct that the user would have to wait for garbage collection to kick in. A quick-fix would be to cache by username+userpass but I'm concerned that would lead to logging the password.

I'll test but perhaps caching by something else is preferred.

Reply:
maybe cache by `<username>-<hash(password)>`?

Reply:
Spoke with @BenPope about this.
1. No need to have an external hash because the container already does that
2. If the user changes their password, then we'll also need to change the config on the client side.

An approach for 2:
1. Check if user password changed.
2. If yes, store the old client away for later, make a new client with the new password and attach it to the user, stop the old client and remove it from the container.